src/main/java/hudson/plugins/mercurial/MercurialSCMSource.java

```diff
@@ -180,21 +173,21 @@ public void setCredentialsId(@CheckForNull String credentialsId) {
     }
 
     @Deprecated @Restricted(DoNotUse.class) @RestrictedSince("2.0") public HgBrowser getBrowser() {
-        for (SCMSourceTrait t: traits) {
-            if (t instanceof MercurialBrowserSCMSourceTrait) {
-                return ((MercurialBrowserSCMSourceTrait) t).getBrowser();
-            }
-        }
-        return null;
+        MercurialBrowserSCMSourceTrait t = getTrait(MercurialBrowserSCMSourceTrait.class);
+        return t != null ? t.getBrowser() : null;
     }
 
     @Deprecated @Restricted(DoNotUse.class) @RestrictedSince("2.0") public boolean isClean() {
+        return getTrait(CleanMercurialSCMSourceTrait.class) != null;
+    }
+
+    private @CheckForNull <T extends SCMSourceTrait> T getTrait(@Nonnull Class<T> traitClass) {
```

Comment:
I am to take it that it is illegal to have multiple instances of the same trait class? I guess you are using the corresponding control mode in `scm-api`?